README.md
# RPG-Pseudocode - Inspired by the Game Above and Below
#Rewards, payments, and end-game points may or may not change depending on the difficulty of the game

"""Currencies and objects:
You lose the game if you are out of health
Gold (to buy or sell items, you start the game with one gold in your inventory.)
Attack (to be able to complete quests)
Weapons: rusty knife, pocket knife, dagger, broadsword, rusty broadsword, healing sword (when you earn a new weapon, it adds to the total amount of attack you have)
Reputation (increases or decreases prizes)
Health (to complete quests if you don't have enough attack, player starts with two health. You cannot complete a quest if you do not have enough health)
Potions (heals one health)
Elixir (increases max health by 1)
"""

"""When you fail the game: goes to a menu disaplaying, "Game Over" with a restart option."""


"""A thin woman with a wooden staff and dark cloak approaches you, her face hidden in the shoadow of her hood. "I must ask something of you," she says. "My traveling companions betrayed me and stole a precious tome of knowledge.  It tells the legend of an ancient kingdom that ruled these lands long ago, and is the last of its kind. Could you please help me retrieve it?"

  If you help the woman: you earn one reputation. The woman joins your party. She's very frail, but she will do her best to help. Gain two health, two more attack, and four gold
  If you refuse: you lose one reputation
"""
  
"""You enter a small village, and a merchant wants to sell you an item. It will cost you one gold, and it doesn't look like it's in good condition. He tells you it would do you good in the long run, but it looks like an old letter-opener...
   If the player chooses to buy the item: gain a reputation and a rusty knife to inventory (attack: 1) (removes one gold from currency)
   If the player refuses to buy the item: the merchant looks a little disappointed but thanks you for your time.
"""
 
"""On your way, you must cross a smooth, clear river. You decide to board and see what you can find further downstream. You pass forests of huge mushrooms, bats hanging in silent swarms in tall chambers, and even a few outposts along the rocky bank. As you turn into a new passage, the water grows rough. Sharp, deadly rocks jut up in every direction, threatening to cut your little boat to pieces. There's a sandy shore on your right.
  If you stop the boat and explore: A potion if you have less than one reputation, a potion and two gold if you have one reputation.
  If you brave the rapids: You lose two health, but gain six gold and a reputation.

"""You venture into a massive cavern filled with the ruins of a crumbling city. As you wander among the cracked stone walls you smell burning wood, and soon come upon a campfire, a cloaked traveler sitting behind it. "I've explored every inch of this city, and you'll find nothing here. But, for a price, perhaps I could be persuaded to part with my knowledge of this place." For all you know, he may have just arrived.
  If you convince the man to tell you about the city: pay one gold and gain one reputation, or pay two gold and gain two reputation. The man tells you that there is great evil nearby, and that you should be careful. You are confused by his words, since there is nothing nearby, but it doesn't hurt to be extra cautious.
  If you explore the city yourself: you find one potion, nothing much else around here.
"""

"""Your party finds a tent set up underneath a massive face cut into the cave rock. The embers of a campfire still glow, and you assume the owner of the tent is nearby. It might be one of the betrayers the frail woman mentioned, but it is a squat, grey-skinned creature with round, red eyes and rows of spikes on its pack - it's a glogo.
  Loot his tent before he gets back: -2 reputation, three gold, one pocket knife (attack +1)
  Befriend the glogo: Glogo joins the party, max health +1 and attack +3
"""

"""You stop at a trading post. Bright lanterns hang on the dock, and little boats fastened with thick rope bob up and down in the water.
  You can buy a dagger for four gold and attack increases by three, or you can buy potions for two gold each.

--> If the frail woman hasn't joined your party: You see the same frail woman again, she smiles, and you end up feeling really sorry for her. You agree to help her find her tomb of knowledge, even though you seem like it's a waste of time.
  """
  
"""Your party travels through a swamp and soon you come upon a fishfolk woman. "Can you help me find mushrooms?" she asks. "My village is running out of food and we need help." Do you help her find mushrooms or ignore her and continue exploring?
   Search for mushrooms: you earn one reputation and 5 gold, you can lose one health and gain two reputation and ten gold.
   Continue exploring: you lose one reputation and one health if you have attack 3, or you lose one reputation and health but gain 5 gold if you have attack 6.
"""

"""You turn a corner and come across a hug pile of bones. Moments later, a massive, blue cat emerges from the darkness. The cat's eyes are thin slits of purple, its striped, bushy tail swaying back and forth. "Find us some fish," says the cat, eyeing you hungrily. Do you fight the cat, hide from it, or search for some fish to give it?
   Hide: you lose one reputation
   Search for some fish for the cat: the cat joins your party, and your attack increases by 4.
   Fight the cat: requires attack 6, you lose two reputation and gain five gold
"""

"""Your party follows a wide passage into a valley when you come upon many bright tents of red and yellow. Tall torches and lanterns light up what seems to be a market. As you peruse the wares offered by creatures of every shape and size, a red-bearded man leaning on a gnarled staff approaches you. "I have a proposition that will be well worth your time, if you'd care to listen," he says. Do you listen to him or haggle with the merchants?
   Haggle: Requires reputation 2, and you can pay twenty gold if you have the rusty knife and gain a healing sword. Attack +5 and you gain back a health every time you fight. Otherwise, you can buy potions for 2 gold or an elixir for 5 gold.
   Listen to the man's proposition:
"I can tell you more when we reach my camp," says the bearded man. He beckons you to follow and leads you away from the circle of merchant tents, down a few dark passages, stopping in a round cave. You don't see any sign of his camp. Suddenly, five men emerge from the shadows, armed with heavy clubs and rocks. "My proposition is this," says the man with a wicked grin. Empty your pockets and we won't hurt you." Do you pay the thugs or fight them?
   Pay: -1 reputation, -10 gold, you may lose a weapon or two if you don't have enough gold, or you lose your health if you don't have enough weapons.
      (one weapon would be sold for half it's value, rounded down, and a health would get you one gold. If you don't have enough health, you lose the game.)
   Fight: If you have attack 5: 2 coins, +1 reputation. If you have attack 10: 4 coins, +2 reputation
"""
   
"""After crawling up to the side of a steep cliff, your party sees two men running in the distance. "Those were my traveling companions!" yells the woman. It doesn't seem likely you'll catch up to them.
   Attempt to chase them: requires attack 7, you aren't able to catch them, but they did drop a bag containing ten gold and one potion.
   Explore the area to find a shortcut: There are no shortcuts, but you can still see those two men running away. You find a rusty broadsword (attack +3) and an elixir.
"""

"""You come across a traveling merchant dressed in yellow and reining in an old, green-skinned pack-lizard. "Come, take a look at my wares," says the merchant as shafts of daylight slice through the ceiling and light up his veiled face. The man is traveling alone, and as far as you can tell, he is unarmed. You could probably snatch one of his bags and run before he would be able to stop you. Do you haggle with the man, or steal from him?
   Haggle: You can buy a potion for 1 gold, an elixir for five gold, or a broadsword for 10 gold (attack +5)
   Steal from the merchant: requires reputation below zero and you lose three reputation, gain seven coins, and gain three potions.
"""

"""Attempting to follow the woman's betrayers, you arrive at apassage blocked by a huge scorpion. Behind the scorpion, you can still see them running. How the heck did they get past this thing?"""
   Fight the scorpion: Attack 15, gain one reputation
   Attempt to flee the scorpion: Lose three health
"""

"""You catch up to the two men and corner them. The frail woman smiles warmly. "Thank you for retrieving my ancient tome..." She laughs wickedly and transform into a dark, shadowy demon with flaming eyes. "I'm finally free after hundreds of years! Thank you, foolish creatures. Now I will destroy you," shrieks the demon. You must fight the demon, and it requires attack 20. (Frail woman leaves your party and you lose 2 health and 2 attack if she joined at the beginning of the game -> may have to change this mechanic if the coding is too difficult.)

If you succeed:
You defeat the demon, and it shrieks while disappearing into a cloud of smoke. At least the frail woman wasn't lying about the tome. You use the tome to find the ruins of a great civilization, and hopfully one day, it can be habitable once again.
"""

"""END GAME POINTS:
Max health - one point each
Weapons - rusty knife (1 point) , pocket knife (1 point), dagger (2 points), broadsword (3 points), rusty broadsword (2 points), healing sword (5 points)
Healing items - elixir (1 point)
Reputation - -2 points if you have less that -3 reputation, or 2 points if you have more than 3 reputation
Gold - 1 point for every 5 gold

If the player has more than 25 points, gets a secret ending:
You help rebuild the ancient civilization, and there are remenants of ancient technology used to repel monsters and evil. Hopefully one day, you can use this to make the world a better and safer place.
"""


Changelog.md
==============================================================

Continuous-Game-Play.py

v1.0:
- This is the first version

v1.1:

- Added paramaters to heal_health and modify-gold
- Changed visual indentation to the python style guide
- Added spaces around operators

v1.2:

-Pushed (this) changelog from Replit
-Fixed errors in Python Style Guide

==============================================================

RPG-Inventory.py

v1.0:

- This is the first version

v1.2:

- Added printing functions to print dictionaries nicely
- Bug fixes

v1.3:

- Changed inventory to nested list with dictionary
- Fixed add_characters function
- Added spaces around operators
- Fixed potion function (bug fixes)

v1.4:

- Made changes to follow the python style guide
- Bug fixes

v1.5:
- More bug fixes
<<<<<<< HEAD

==============================================================

RPG-Modules-and-Maps.py

v1.0:

- Separated RPG-Inventory.py into different modules
- This is the first version

v1.1:

- Added Maps.py
- Bug fixes
- Defined player location

v1.2:

- Added alternate lists to insert into map to show player's location
- Removed circular imports

v1.3:
- Bug fixes related to the game map
- Attempted to remove duplicate "You're here" tiles on map

v1.4:
- Removed duplicate "You're here" tiles 
- Made the Healing sword not available in market until the rusty knife is purchased.

v1.5:
- Added purchase functions with paramaters to make buy_items() neater
- Bug fixes
- Checked PEP8
- Fixed recursion issue

v1.6 (Different assignment):
- Added exception handling
- Bug fixes

v1.7 (Different assignment):
- Changed weapons and items into classes
- Bug fixes
- Checked python style

v1.8
- Added monster classes
- Bug fixes

====================================================================
>>>>>>> origin/main


Map.py
# Catherine McLellan
# Map and player position
# CS 30
# Oct. 20, 2021
import Locations
import random

row = 2
column = 2
# So that the map doesn't print out when the user doesn't ask for it
print_command = 0


class Map:
    """Generating and printing a map"""
    def __init__(self, map_tiles):
        self.map_tiles = map_tiles

    def __str__(self):
        return self.map_tiles


def update_map():
    game_map.map_tiles = full_map
    print(game_map)

# location_list2 will be reevaluated as this list
# There are extra spaces so the game map looks nice
location_list = [
    ["   Monster  ", "   Market   ", "    Gold    ",
        "    Witch   ", "  New ally  "],
    ["    Gold    ", "    Gold    ", "    Gold    ",
        "   Monster  ", "    Witch   "],
    ["    Gold    ", "   Monster  ", "    Start   ",
        "    Gold    ", "   Market   "],
    ["   Market   ", "    Witch   ", "    Gold    ",
        "   Monster  ", "   Monster  "],
    ["    Gold    ", "   Market   ", "    Gold    ",
        "   Monster  ", "    Gold    "]
]

# Making a list with the same values that get replaced by "You're here"
# so the "You're here" tiles don't overwrite the game map
location_list2 = [
    ["   Monster  ", "   Market   ", "    Gold    ",
        "    Witch   ", "  New ally  "],
    ["    Gold    ", "    Gold    ", "    Gold    ",
        "   Monster  ", "    Witch   "],
    ["    Gold    ", "   Monster  ", "    Start   ",
        "    Gold    ", "   Market   "],
    ["   Market   ", "    Witch   ", "    Gold    ",
        "   Monster  ", "   Monster  "],
    ["    Gold    ", "   Market   ", "    Gold    ",
        "   Monster  ", "    Gold    "]
]

full_map = """
+--------------------------------------------------------------------------+
| {}""".format(location_list2[0][0]) + " | {}".format(location_list2[0][1]) + " | {}".format(location_list2[0][2]) + " | {}".format(location_list2[0][3]) + " | {} |".format(location_list2[0][4]) + """
+--------------------------------------------------------------------------+
| {}""".format(location_list2[1][0]) + " | {}".format(location_list2[1][1]) + " | {}".format(location_list2[1][2]) + " | {}".format(location_list2[1][3]) + " | {} |".format(location_list2[1][4]) + """
+--------------------------------------------------------------------------+
| {}""".format(location_list2[2][0]) + " | {}".format(location_list2[2][1]) + " | {}".format(location_list2[2][2]) + " | {}".format(location_list2[2][3]) + " | {} |".format(location_list2[2][4]) + """
+--------------------------------------------------------------------------+
| {}""".format(location_list2[3][0]) + " | {}".format(location_list2[3][1]) + " | {}".format(location_list2[3][2]) + " | {}".format(location_list2[3][3]) + " | {} |".format(location_list2[3][4]) + """
+--------------------------------------------------------------------------+
| {}""".format(location_list2[4][0]) + " | {}".format(location_list2[4][1]) + " | {}".format(location_list2[4][2]) + " | {}".format(location_list2[4][3]) + " | {} |".format(location_list2[4][4]) + """
+--------------------------------------------------------------------------+"""


game_map = Map(full_map)


def reset_location_list():
    """Resetting the "You're here" tiles to the original location_list"""
    global row
    global column
    if location_list2[row][column] == " You're here":
        location_list2[row][column] = location_list[row][column]


def player_location():
    """Calls functions based on the tile the player is located at"""
    if location_list[row][column] == "    Start   ":
        print("You are at the start. Please choose an action.")
    elif location_list[row][column] == "    Gold    ":
        Locations.modify_gold(random.choice(range(1, 4)))
    elif location_list[row][column] == "   Monster  ":
        print(Locations.locations["Monster"])
        Locations.fight_monster()
    elif location_list[row][column] == "    Witch   ":
        Locations.witch()
    elif location_list[row][column] == "   Market   ":
        print(Locations.locations["Market"])
        Locations.buy_items()
    elif location_list[row][column] == "  New ally  ":
        Locations.new_ally()


def move_choice():
    """Changes player's position on map depending on their choice"""
    global row
    global column
    global print_command
    global location_list2
    reset_location_list()
    valid_input = False
    # Repeats loop until user puts in a valid answer
    while not valid_input:
        direction = input("Do you want to go north, south, east, or west?")
        try:
            if direction == "north":
                row = row - 1
            elif direction == "south":
                row = row + 1
            elif direction == "east":
                column = column + 1
            elif direction == "west":
                column = column - 1
            else:
                print("Invalid input!")
                continue
            if row < 0:
                row = row + 1
                raise IndexError
            elif row > 4:
                row = row - 1
                raise IndexError
            elif column < 0:
                column = column + 1
                raise IndexError
            elif column > 4:
                column = column - 1
                raise IndexError
        except IndexError:
            print("You can't go in this direction!")
        else:
            print_command = 1
            player_location()
            location_list2[row][column] = " You're here"
            print_command = 0
            valid_input = True


Gameplay.py
# Catherine McLellan
# Gameplay function
# CS 30
# Oct. 20, 2021
import Inventory
import Map
options = "1. Explore", "2. See inventory", "3. Heal", "4. See map", "5. Quit"


def gameplay():
    """Funciton that starts gameplay and inputs player choices"""
    while True:
        for choice in options:
            print(choice)
        player_choice = input("Which action do you wish to choose?")
        if player_choice == "1":
            Map.move_choice()
        elif player_choice == "2":
            Inventory.see_inventory()
        elif player_choice == "3":
            Inventory.use_potions()
        elif player_choice == "4":
            Map.update_game_map()
        elif player_choice == "5":
            print("Thanks for playing!")
            break
        else:
            print("""Invalid input!
Please press a number for your chosen action.""")
        if Inventory.health <= 0:
            print("You're out of health!")
            print("GAME OVER")
            break


RPG-Inventory.py
# CS 30 RPG Inventory Assignment
# Catherine McLellan
# October 5, 2021
import random
# Starting inventory for player:
inventory = {"potions": 0, "elixirs": 0, "weapons": []}
party = ["Yourself"]

# Player stats at beginning of the game:
health = 2
max_health = 2
attack = 0
gold = 0

# Listing weapons, healing items, characters, and locations in the game:
weapon = {
    "Rusty knife":
    {
        "Description: ": "A puny knife that has rusted over.",
        "Price: ": "Costs 1 gold"},
    "Pocket knife":
    {
        "Description: ": "A tiny blade. Looks like a letter opener.",
        "Price:": "Costs 2 gold"},
    "Dagger":
    {
        "Description: ": "Slightly better than a pocket knife.",
        "Price: ": "Costs 5 gold"},
    "Healing sword":
    {
        "Description: ": "Upgraded rusty knife, heals you when you fight!",
        "Price: ": "Costs 20 gold and requires the rusty knife."},
    "Rusty broadsword":
    {
        "Description: ": "Looks really old but still has some fight left.",
        "Price: ": "Costs 8 gold"},
    "Broadsword":
    {
        "Description: ": "A hero's sword.",
        "Price: ": "Costs 15 gold"}
    }

healing_stuff = {
    "Potion":
    {"Description: ": "Heals one health.",
        "Cost: ": "5 gold."},
    "Elixir":
    {"Description: ": "Adds to maximum health.",
        "Cost: ": "15 gold."}
}


characters = {
    "Frail Woman":
    {
        "Description: ": "She's very frail, but she will do her best to help.",
        "Stats: ": "Attack increases by 2, Max health increases by 2"},
    "Glogo":
    {
        "Description: ":
        """A squat, grey-skinned creature with round, red eyes,
and rows of spikes on its back.
        """,
        "Stats: ": "Attack increases by 3, Max health increases by 1"},
    "Cat":
    {
        "Description: ":
        "A massive, blue cat that looks like it wants to eat you.",
        "Stats: ": "Attack increases by 4"}
}

locations = {
    "Gold": "You found some gold!",
    "Monster": "Fight a goblin!",
    "Witch": "Witch steals a health!",
    "Market": "Purchase weapons or healing items!",
    "New ally!": "New character joins party!"
}
player_options = "1. Explore", "2. See inventory", "3. Heal", "4. Quit"


def see_inventory():
    """Displays inventory, party, and player status"""
    if inventory == [[0, 0], []]:
        print("You don't have any healing items or weapons!")
    elif inventory != [[0, 0], []]:
        total_potions = f"""You have {inventory["potions"]} potions and """
        total_elixirs = f"""{inventory["elixirs"]} elixirs."""
        if inventory["potions"] == 1:
            total_potions = f"""You have {inventory["potions"]} potion and """
        if inventory["elixirs"] == 1:
            total_elixirs = f"""{inventory["elixirs"]} elixir."""
        print(total_potions + total_elixirs)
        if inventory["weapons"] == []:
            print("""You have no weapons!""")
        for weapon in inventory["weapons"]:
            print(f"""You have a {weapon}.""")
    print("""Your party: """ + ", ".join(map(str, party)))
    print("Your total attack is " + str(attack) + ".")
    print("You have " + str(gold) + " gold.")
    print("You have " + str(health) + "/" + str(max_health) + " health.")


def use_potions():
    """To increase / heal health by 1 if the player has enough items"""
    global health
    global max_health
    heal_item_to_use = input("""
    Do you want to use a potion, an elixir, or continue exploring?""")
    if heal_item_to_use == "potion":
        if health < max_health and inventory["potions"] > 0:
            inventory["potions"] = inventory["potions"] - 1
            health = health + 1
            print("You healed one health!")
        elif inventory["potions"] == 0:
            print("You don't have any potions!")
        elif health == max_health:
            print("You can't heal!")
    elif heal_item_to_use == "elixir":
        if inventory["elixirs"] > 0:
            max_health = max_health + 1
            inventory["elixirs"] = inventory["elixirs"] - 1
            print("Your max health has increased by 1!")
        elif inventory["elixirs"] == 0:
            print("You don't have any elixirs!")
    elif heal_item_to_use == "explore":
        gameplay()
    else:
        print("""Invalid input!
Please type 'potion', 'elixir', or 'explore'.""")
        use_potions()


def modify_gold(amount_gold):
    """Modifies player's gold"""
    global gold
    gold = gold + amount_gold
    if amount_gold >= 1:
        print(f"""You found {amount_gold} gold!""")


def fight_goblin():
    """Fight a monster, game quits if you fail"""
    global attack
    global health
    if "healing sword" in inventory["weapons"]:
        if health < max_health:
            health = health + 1
    goblin_health = 4 - attack
    if goblin_health <= 0:
        modify_gold(4)
        print("You defeated the goblin and earned four gold!")
    elif goblin_health > 0:
        goblin_outcome()


def goblin_outcome():
    """If the player doesn't have enough attack to defeat the goblin"""
    global gold
    goblin_health = 4 - attack
    fight_or_flight = input("""You don't have enough attack!
Do you want to flee or take damage to defeat the goblin?
(type 'flee' or 'take damage')""")
    if fight_or_flight == "flee":
        gold_stolen = random.choice(range(1, 4))
        if gold >= gold_stolen:
            gold_stolen = 0 - gold_stolen
            modify_gold(gold_stolen)
            print("The goblin stole some gold!")
        print("You ran away from the goblin!")
    elif fight_or_flight == "take damage":
        # Using your health to defeat the goblin
        global health
        health = health - goblin_health
        if health > 0:
            modify_gold(5)
            print("You defeated the goblin and earned five gold!")
            # If you try to use your health but don't have enough:
        elif health <= 0:
            print("You can't beat the goblin!")
    else:
        print("Please type 'flee' or 'take damage'.")
        goblin_outcome()


def witch():
    """Witch decreases max health by 1"""
    print(locations["Witch"])
    global max_health
    global health
    if health == max_health:
        health = health - 1
    max_health = max_health - 1
    return max_health


def print_organized(item, indenting=-5):
    """Organizing print for nested dictionaries"""
    if type(item) == dict:
        print('')
        indenting += 5
        for attributes in item:
            print(indenting * ' ', end='')
            print(attributes, end='')
            print_organized(item[attributes], indenting)
    else:
        print(item)


def buy_items():
    """Gives the player the option to purchase an item."""
    print("Market Inventory:")
    print_organized(weapon)
    print_organized(healing_stuff)
    print("")
    choice = input("Would you like to make a purchase?")
    if choice == "yes":
        global gold
        global attack
        purchase_choice = input("What would you like to purchase?")
        if purchase_choice in inventory["weapons"]:
            print("Please select a weapon we have in stock!")
            buy_items()
        if purchase_choice == "rusty knife":
            if "rusty knife" not in inventory["weapons"]:
                if "healing sword" in inventory["weapons"]:
                    print("Please select something we have in stock!")
                    buy_items()
                if gold >= 1:
                    gold = gold - 1
                    inventory["weapons"].append("rusty knife")
                    attack = attack + 1
                    del(weapon["Rusty knife"])
                    print("Thank you for your purchase!")
                elif gold < 1:
                    print("You don't have enough gold!")
        elif purchase_choice == "pocket knife":
            if "pocket knife" not in inventory["weapons"]:
                if gold >= 2:
                    gold = gold - 2
                    inventory["weapons"].append("pocket knife")
                    attack = attack + 1
                    del(weapon["Pocket knife"])
                    print("Thank you for your purchase!")
                elif gold < 2:
                    print("You don't have enough gold!")
        elif purchase_choice == "dagger":
            if "dagger" not in inventory["weapons"]:
                if gold >= 5:
                    gold = gold - 5
                    inventory["weapons"].append("dagger")
                    attack = attack + 3
                    del(weapon["Dagger"])
                    print("Thank you for your purchase!")
                elif gold < 5:
                    print("You don't have enough gold!")
        elif purchase_choice == "healing sword":
            if ("healing sword" not in inventory["weapons"] and
               "rusty knife" in inventory["weapons"]):
                if gold >= 20:
                    gold = gold - 20
                    inventory["weapons"].append("healing sword")
                    attack = attack + 5
                    del(weapon["Healing sword"])
                    inventory["weapons"].remove("rusty knife")
                    print("Thank you for your purchase!")
                elif gold < 20:
                    print("You don't have enough gold!")
            else:
                print("You can't purchase this weapon!")
        elif purchase_choice == "rusty broadsword":
            if "rusty broadsword" not in inventory["weapons"]:
                if gold >= 8:
                    gold = gold - 8
                    inventory["weapons"].append("rusty broadsword")
                    attack = attack + 2
                    del(weapon["Rusty broadsword"])
                    print("Thank you for your purchase!")
                elif gold < 5:
                    print("You don't have enough gold!")
        elif purchase_choice == "broadsword":
            if "broadsword" not in inventory["weapons"]:
                if gold >= 15:
                    gold = gold - 15
                    inventory["weapons"].append("broadsword")
                    attack = attack + 4
                    del(weapon["Broadsword"])
                    print("Thank you for your purchase!")
                elif gold < 15:
                    print("You don't have enough gold!")
        elif purchase_choice == "potion":
            if gold >= 5:
                inventory.update({"potions": inventory["potions"] + 1})
                gold = gold - 5
                print("Thank you for your purchase!")
            elif gold < 5:
                print("You don't have enough gold!")
        elif purchase_choice == "elixir":
            if gold >= 15:
                inventory.update({"elixirs": inventory["elixirs"] + 1})
                gold = gold - 15
                print("Thank you for your purchase!")
            elif gold < 15:
                print("You don't have enough gold!")
        else:
            print("I don't understand your request!")
            buy_items()
    elif choice == "no":
        print("Thank you for your time!")
    else:
        print("Not a valid answer! Please type 'yes' or 'no'.")
        buy_items()


def characters_organized(d, indent=0):
    """Organizing character lists"""
    s = ""
    for key, value in d.items():
        s += '\t' * indent + str(key)
        if isinstance(value, dict):
            characters_organized(value, indent+1)
        else:
            s += '\t' * (indent+1) + str(value) + "\n"
    return s


def new_ally():
    """A new ally will join the party (in a certain order)"""
    global attack
    global max_health
    if "Frail Woman" not in party:
        party.append("Frail Woman")
        print("A frail woman has joined your party!\n" +
              characters_organized(characters["Frail Woman"]))
        attack = attack + 2
        max_health = max_health + 2
    elif "Glogo" not in party:
        party.append("Glogo")
        print("A glogo has joined your party!\n" +
              characters_organized(characters["Glogo"]))
        attack = attack + 3
        max_health = max_health + 1
    elif "Cat" not in party:
        party.append("Cat")
        print("A cat has joined your party!\n" +
              characters_organized(characters["Cat"]))
        attack = attack + 4
    elif party == ["Yourself", "Frail Woman", "Glogo", "Cat"]:
        print("All of the characters have joined your party!")


def choose_random_location():
    """Generates a random location with a weighted random choice."""
    random_location = random.choices(list(locations),
                                     weights=[50, 20, 7, 20, 3], k=1)
    if random_location == ["Gold"]:
        modify_gold(random.choice(range(1, 4)))
    elif random_location == ["Monster"]:
        print(locations["Monster"])
        fight_goblin()
    elif random_location == ["Witch"]:
        witch()
    elif random_location == ["Market"]:
        print(locations["Market"])
        buy_items()
    elif random_location == ["New ally!"]:
        new_ally()


def gameplay():
    """Funciton that starts gameplay and inputs player choices"""
    while True:
        for choice in player_options:
            print(choice)
        player_choice = input("Which action do you wish to choose?")
        if player_choice == "1":
            choose_random_location()
        elif player_choice == "2":
            see_inventory()
        elif player_choice == "3":
            use_potions()
        elif player_choice == "4":
            print("Thanks for playing!")
            break
            break
        else:
            print("""Invalid input!
Please press 1 to explore, 2 to see your inventory, or 3 to quit.""")
        if health <= 0:
            print("You're out of health!")
            print("GAME OVER")
            break

gameplay()
